Constant folding during path checking

msooseth opened the pull request:
## Description
This is a PR I have been sitting on, and had some partial experience regarding its performance, but I finally got some better proof :) While running PR #352 , I noticed:

```
Checking file: /nix/store/sayhv9lb1nl58rrwsgx928bcng20hqd1-source/test/libyul/yulOptimizerTests/unusedAssignEliminator/for_multi_break.yul

WARNING: hevm was only able to partially explore the given contract due to the following issues:

  - Max Iterations Reached in contract: SymAddr "entrypoint" pc: 19

Found 720801 total pairs of endstates
Reuse of previous queries was Useful in 364751 cases
"OK. Took 154.138271977s seconds"
```

Which is sad. On `main`, it behaves exactly the same. Sad. I thought... maybe my PR can fix it! And indeed it can:

```
Checking file: /nix/store/sayhv9lb1nl58rrwsgx928bcng20hqd1-source/test/libyul/yulOptimizerTests/unusedAssignEliminator/for_multi_break.yul

WARNING: hevm was only able to partially explore the given contract due to the following issues:

  - Max Iterations Reached in contract: symbolic(entrypoint) pc: 19

Found 37249 total pairs of endstates
Asking the SMT solver for 16900 pairs
Reuse of previous queries was Useful in 15286 cases
"OK. Took 7.409072883s seconds"
OK (7.42s)
```

That's some nice improvement right there! So now I'm more confident in the performance of this PR :) This PR basically adds a certain level of constant folding to branch exploration. It's a slightly enhanced version of #329 .


## Checklist

- [x] tested locally
- [ ] added automated tests
- [ ] updated the docs
- [x] updated the changelog

d-xo commented on src/EVM/Expr.hs:
```diff
@@ -1057,3 +1058,52 @@ evalProp prop =
+  where
+    constProp :: Prop -> State ConstState Prop
+    constProp prop = go (evalProp prop)
+    go :: Prop -> State ConstState Prop
```
Why do we need the return value here? Don't we throw it away with the call to `execState` later?

msooseth replied:
Yeah, good point. This has now been fixed!

d-xo replied:
we can still simplify further here and change the type of `go` to `Prop -> State ConstState ()`

msooseth replied:
Ah yess! I finally understood what you meant, thanks! OK, it's done! I'm resolving now. Can you please check and then we can merge!

d-xo commented on src/EVM/Expr.hs:
```diff
@@ -1057,3 +1058,52 @@ evalProp prop =
+
+-- | Folds constants
+constFoldProp :: [Prop] -> ConstState
+constFoldProp ps = execState (mapM constProp ps) (ConstState mempty True)
```
```suggestion
constFoldProp ps = execState (mapM (go . evalProp) ps) (ConstState mempty True)
```

d-xo commented on src/EVM/Expr.hs:
```diff
@@ -1057,3 +1058,52 @@ evalProp prop =
+  deriving (Show)
+
+-- | Folds constants
+constFoldProp :: [Prop] -> ConstState
```
I think we could just return `canBeSat` here?

msooseth replied:
Oh, yes, of course! Good idea :) Fixed in 88381efaaa93df187691d2cb404d8c8598ec5722

d-xo commented on src/EVM/SymExec.hs:
```diff
@@ -308,9 +311,12 @@ interpret fetcher maxIter askSmtIters heuristic vm =
                     performQuery
                   -- otherwise just try both branches and don't ask the solver
```
can we update the comment here?

msooseth replied:
Ooops, thanks, good point! Fixed in 2d29afa4542dedf66ac3afe9863580695b89cfad

msooseth commented:
OK, so this should now be ready to merge I think? I need one more review for the fix to the `constFoldProp`'s return value being removed and we'll be good to go!! Yaaaaay

d-xo commented on src/EVM/Expr.hs:
```diff
@@ -845,6 +846,8 @@ simplify e = if (mapExpr go e == e)
 simplifyProps :: [Prop] -> [Prop]
 simplifyProps = remRedundantProps . map evalProp . flattenProps
 
+
+
```
```suggestion
```

msooseth commented:
I realized we need a slightly more powerful prop simplifier, so I added a subfuction, `simpInnerExpr`.

I also realized we introduced a subtle "bug" that made some of our tests not very useful: We added simplification to `assertProps`, but we were using it to check simplification :D So now I added a function, `assertPropsNoSimp` that is called from `assertProps`, after simplification. This way, we can call it to check simplification correctness.

I also added a bunch of concrete tests, which highlighted the need for `simpInnerExpr`. This means this PR need a new review.... I'm sorry. I hope these new changes are not too big and make sense.
